controller/controller_server/csi_controller_server.py

```diff
@@ -400,9 +398,41 @@ def ControllerUnpublishVolume(self, request, context):
 
     def ValidateVolumeCapabilities(self, request, context):
         logger.info("ValidateVolumeCapabilities")
-        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
-        logger.info("finished ValidateVolumeCapabilities")
-        return csi_pb2.ValidateVolumeCapabilitiesResponse()
+        try:
+            utils.validate_validate_volume_capabilities_request(request)
+
+            secrets = request.secrets
+            user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+            array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+            space_efficiency = request.parameters.get(config.PARAMETERS_SPACE_EFFICIENCY)
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                array_mediator.validate_supported_space_efficiency(space_efficiency)
+
+                volume = array_mediator.get_object_by_id(object_id=volume_id, object_type=config.VOLUME_TYPE_NAME)
+
+            if not volume:
+                raise controller_errors.ObjectNotFoundError(volume_id)
+
+            if request.volume_context:
+                utils.validate_volume_context_match_volume(request.volume_context, volume)
+
+            utils.validate_parameters_match_volume(request.parameters, volume)
+
+            logger.info("finished ValidateVolumeCapabilities")
+            return utils.generate_csi_validate_volume_capabilities_response(request.volume_context,
+                                                                            request.volume_capabilities,
+                                                                            request.parameters)
+        except (controller_errors.ObjectNotFoundError, ObjectIdError) as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+            return csi_pb2.ValidateVolumeCapabilitiesResponse(message=ex.message)
+        except (ValidationException, controller_errors.SpaceEfficiencyNotSupported) as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            return csi_pb2.ValidateVolumeCapabilitiesResponse(message=ex.message)
 
```

Comment:
according to the [spec](https://github.com/container-storage-interface/spec/blob/v1.3.0/spec.md#validatevolumecapabilities-errors):
> If the plugin is unable to complete the ValidateVolumeCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.

so what about `except Exception`? I know, pokemon exception handling, but that's currently the convention in this project

Reply:
you right..

Reply:
ok, I caught them all..